host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,430 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+					  key_len) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, secret_len, ==, key_len) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_SECRET_KEY) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_key_type, ==,
+					  CKK_GENERIC_SECRET) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_local, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_sensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_extract, ==, CK_TRUE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_asensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_nextract, ==, CK_FALSE))
+		goto out;
+
+	rv = C_DestroyObject(session, derv_key_hdl);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Invalid parameters during derivation");
+
+	/* Length of data used for derivation < key length */
+	data_len = 16;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Data is not multiple of 16 */
+	data_len = 18;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Wrong Mechanism */
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_MECHANISM_INVALID, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if operation already active");
+
+	/* Generate an AES key which can perform Encryption */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template_w_enc_der,
+			   ARRAY_SIZE(parent_template_w_enc_der),
+			   &aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_EncryptInit(session, &cktest_aes_cbc_mechanism, aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	data_len = 32;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key2, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_OPERATION_ACTIVE, rv))
+		goto out;
+
+	rv = C_EncryptFinal(session, NULL, NULL);
+	/* Only check that the operation is no more active */
+	if (!ADBG_EXPECT_TRUE(c, rv != CKR_BUFFER_TOO_SMALL))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if parent key C_ENCRYPT is TRUE");
+
+	data_len = 32;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key_enc, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_FUNCTION_FAILED, rv))
```

Comment:
Maybe this test should relax the possible error code values.
Or maybe an inline comment `/* Note strictly expecting FUNCTION_FAILED but expecting a failure */`.

Reply:
Sure, I will add the comment.

Reply:
Needs comment to clarify what is happening in here and why we get `CKR_FUNCTION_FAILED`.

Reply:
Done